### PR TITLE
ci,crusher: increase supported num branches

### PR DESCRIPTION
**Pull request by vicentebolea**
This is needed for the generate_pipelines.py to support processing more than 100 branches in the Gitlab CI (The max per_page items that Gitlab CI allow).

I have also change the functional programming style form map/filter to comprehensions list to make the code more pythonic.

### CODE REVIEW

**Comment by vicentebolea** [comment]
@scottwittenburg note that we cannot test this until it gets merged since Crusher runs master's generate_pipelines.py

File: .gitlab/config/generate_pipelines.py [hunk 1]
```diff
@@ -88,14 +127,13 @@ def has_no_status(branch):
...
-
-    # Select the arg.max most least recent branches
+    branches = request_as_list(gl_url + '/repository/branches')
+    branches = [add_timestamp(branch) for branch in branches]
+    branches = [b for b in branches if is_recent(b)]
     branches = sorted(branches, key=lambda x: x['dt'])
-    branches = itertools.islice(branches, args.max)
+
+    run_n_times = skip_after_n_successes(default_value=False, n=args.max)
+    branches = [b for b in branches if run_n_times(has_no_status, b)]
```
Comment (vicentebolea) [inline .gitlab/config/generate_pipelines.py:136]: This is needed to skip making the rest api after we find our requested n number of branches. Gitlab/GitHub rest api have strong limits on how many request we can do per min.

Comment (scottwittenburg) [inline .gitlab/config/generate_pipelines.py:136]: It has taken me some time to figure what the heck this is doing, so I think a descriptive comment would make sense here.  The way I'm reading this code it's intent is to "schedule jobs for the next `args.max` (i.e. 2) branches for which github doesn't have an associated status".  Is that correct though?

Comment (vicentebolea) [reply .gitlab/config/generate_pipelines.py:136]: Sorry I should have made a better introduction :angel: . Yep

File: .gitlab/config/generate_pipelines.py [hunk 2]
```diff
@@ -44,7 +78,12 @@ def has_no_status(branch):
         gh_commit_sha = branch['commit']['parent_ids'][1]
 
     # Query GitHub for the status of this commit
-    commit = request_as_dict(gh_url + '/commits/' + gh_commit_sha + '/status')
+    response = http_get_request(
+        gh_url + '/commits/' + gh_commit_sha + '/status')
+    if int(response.headers['x-ratelimit-remaining']) <= 0:
+        raise ConnectionError(response.json())
```
[the remark this answers is not in the document]
Comment (vicentebolea) [reply .gitlab/config/generate_pipelines.py:84]: Yes, I want to show: hey this error is due to us hitting the github ratelimit

[the remark this answers is not in the document]
Comment (vicentebolea) [reply .gitlab/config/generate_pipelines.py:84]: And perhaps lets the future developer to program this by catching this exception

File: .gitlab/config/generate_pipelines.py (hunk 1 above)
Comment (scottwittenburg) [reply .gitlab/config/generate_pipelines.py:136]: Well, I was suggesting putting a comment in the code so future developers (including me a month from now) get a hint about what's happening.  But I guess that's not strictly necessary.

**Review by scottwittenburg — DISMISSED** [review]

File: .gitlab/config/generate_pipelines.py (hunk 1 above)
Comment (vicentebolea) [reply .gitlab/config/generate_pipelines.py:138]: My bad, find the comment added

File: .gitlab/config/generate_pipelines.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (scottwittenburg) [reply .gitlab/config/generate_pipelines.py:84]: Ok, just curious.

**Review by scottwittenburg — APPROVED** [review]

**Pull request merged**